### ISSUE TITLE
Active unloading

### ISSUE THREAD

**Issue by raxtor** [Stephan-S/FS19_AutoDrive#542]
I was testing a little new feature (WOW).. I know its early version, just my expirience..
- fold/unfold pipe for auger wagon doesnt work
- tractor is turning to combine path after unloading
- some values in GUI settings arent stored correctly or are lost between TAB switching
![fsScreen_2019_09_28_21_25_16](https://user-images.githubusercontent.com/831266/65821584-be949d00-e237-11e9-9ab2-54901059a017.png)
![fsScreen_2019_09_28_21_27_15](https://user-images.githubusercontent.com/831266/65821585-be949d00-e237-11e9-95bf-08019b2e28f9.png)

**Comment by raxtor** [comment]
1.0.5.9 all 3 bugs still there..
![fsScreen_2019_09_30_21_23_38](https://user-images.githubusercontent.com/831266/65909328-aa33da00-e3c8-11e9-8ce3-9d19702ba124.png)

**Comment by raxtor** [comment]
Hi @Stephan-S,
there is something wrong with 1.0.6.0. Pathfinding works very badly :-/ for chasing mode its unusable. In standard mode its a bit better but sometimes unloader also not able find path to combine and ends in circle or combine crash. 
![fsScreen_2019_10_01_22_56_56](https://user-images.githubusercontent.com/831266/66001381-a45df780-e4a1-11e9-85a1-8d70247a7bdc.png)
![fsScreen_2019_10_01_23_09_25](https://user-images.githubusercontent.com/831266/66001548-fa329f80-e4a1-11e9-8ae5-047810ba18c3.png)

I noticed u tried fix fold/unfold pipe.. there are 2 problems:
- pipe is unfolded too late and very often grain cart is missing 1st Wilson bin.. need a few seconds plus
- pipe folding is realized when unloader is going to combine and should be folded when reached waiting point
![fsScreen_2019_10_01_22_58_04](https://user-images.githubusercontent.com/831266/66001699-5695bf00-e4a2-11e9-8690-370f5d2c1e80.png)

sry for troubles.. I noticed also setting "Chase combine" is active for off value. Let me know if I should test this new features with 1 combine only because 1+ has more issues..

**Comment by raxtor** [comment]
Just any idea for circle stuck-- arent you able detect circle stuck somehow.. lets say full left or right riding more than 40 sec.. very often is enough stop/start AD and tractor is able find new path...

**Comment by raxtor** [comment]
@Stephan-S for me it seems there is switched yes/no value for "chase combine" (1.0.6.0), can you check it? Or maybe I can change translation.. but I assume for tag "Chase combine" value yes should chase combine and now it works in oposite way.

**Comment by raxtor** [comment]
Tested 1.0.6.1, it seems also standard mode is corrupted now... tractor is very often in infinite loop and cant find path to combine...

I was thinking about active mode and here is very simplified my approach...I have feel, now unloader has trouble to find right path and trying find correct path again and again and its finished by circle stuck. Also I noticed that combines very often fold/unfold pipes in same time. I am using always more than 1 combine and I am not sure if there is support for multiple machines.
![active_unloading](https://user-images.githubusercontent.com/831266/66167045-4fed8000-e639-11e9-8a11-a4ec65572897.png)
- unloader is waiting for combine pre-call (60% and pipe folded if grain cart)
- when 1 of combine reached 60% of fillLevel than dynamic nav point A is created, unloader is going to A
- new A point position can be moved every specific timeframe if combine is still in move and tractor didnt reach A yet and dynamic path can be improved more to actual position of combine
- in point A tractor check if "Chase combine" is ON and if pipe is over crop
- if pipe over crop, tractor is waiting or chase combine till turn maneuver of CP for fruit over crop
- if pipe is free, tractor is going unload him asap, no need wait for 100% full combine.
- after unloading checks another pre-calls or is going unload or waiting on filed etc.
- if grain cart unfold pipe when tractor left dynamic path to unload nav point

There could be options:
- pre-call level (for 100% is used standard/old mode)
- follow/chase combine (could be removed if unloader life cycle will work as I described upper)
- Stop combine for unloading

**Comment by Stephan-S** [comment]
Thanks for the feedback. Will look into it as soon as I can!

>     * fold/unfold pipe for auger wagon doesnt work



> @Stephan-S for me it seems there is switched yes/no value for "chase combine" (1.0.6.0), can you check it?

The setting is used correct. But if chase combine is disabled and precall is enabled. He will call him over and over again which leads to the driver sort of chasing him, but in a different mode:
Precall means he is using path finder to find a way behind  the combine whereas
chase combine means he is trying to do the active unload part and either follows the combine with some distance and matching speed or drives up to the pipe if there is no fruit detected.



>     * new A point position can be moved every specific timeframe if combine is still in move and tractor didnt reach A yet and dynamic path can be improved more to actual position of combine

This is obviously the part that is causing problems. The rest is basically implemented just as you are describing it here. The problem with this is the computational requirements. I can't do a full path finder call on every frame or even every few seconds. This is a heavy process.
What I am currently working on is a sort of 'quick check' which checks just if the basic straight line between the driver and the combine is collision and fruit free and then uses this path to catch up to the combine. This quickcheck is still a bit expensive and I call this every 5 seconds. But I am still working on making it completely reliable.

>     * if grain cart than unfold pipe when tractor left dynamic path to unload nav point

I will look into the grain cart pipe functions today. They were working as expected in the previous default unloader mode, correct?

Btw.: I saw that you also closed a lot of issues yesterday. Some of them were still somewhat on my radar but I didn't get to it yet. For example the refueling of vehicles. Were you happy with the status of all those closed issues?

**Comment by raxtor** [comment]
> To je samozřejmě ta část, která způsobuje problémy. Zbytek je v podstatě implementován tak, jak to zde popisujete. Problém s tím jsou výpočetní požadavky. Nemohu volat úplný vyhledávač cest v každém snímku nebo dokonce každých pár sekund. To je těžký proces.
V současné době pracuji na jakési „rychlé kontrole“, která kontroluje, zda základní přímka mezi řidičem a kombajnem je bez kolize a bez ovoce, a poté pomocí této cesty dohání kombajn. Tato rychlá kontrola je stále o něco dražší a já tomu říkám každých 5 sekund. Ale stále pracuji na tom, aby byl úplně spolehlivý.

Thx for info how it works. Just a few thoughts.
- for good performance use only static A. If tractor reached A, than make other step. Combines are slow and for small and mid fields it will work smoothly too. Lets say you can check in point A fillLevel and distance of combine and u can make next jump/decission.
- I think you should turn off chase combine for now till pre-call process wont be working without issues. Because its hard for testing and probably we gets mix of issues caused by pre-call or chase without knowledge which one is that. Lets finish pre-call part and than finetuning for chase.
- Grain cart pipe was working well and is working well in 1.0.5.9. Than it stoped working in pre-call mode. After your fixed it, its working somehow but not perfectly (fold/unfold timimng and problem with Wislon trailer 1st bin)
- I closed it because many threads had many old info and I dont want distract you from your targets. I have all in my head. What about locked ToDo/RoadMap thread where you can store ideas which do you want implement? 
Than you dont need let opened old threads full of ballast with 1 good idea. Actually I like you work on active unloading and refueling or so has much lower priority... thats why I closed it.

**Comment by raxtor** [comment]
btw I noticed too 1 version (1.0.6.0?) had reseted front/back offset position and it worked well for 0m offset... I tested only 1 grain cart and combine pipe was in center of trailer. Why it was discarded?

**Comment by raxtor** [comment]
and another question is.. do u need full pathfinder if tractor reached A point? cant you only use mechanism which will move tractor from point A to new position without collision check? Path should be clear.. u can only detect collisons with any object, like other unloader/combine and stop him till path will be clear...

**Comment by Stephan-S** [comment]
No, the combine could have already turned into the next lane or is in the middle of that turning process

**Comment by raxtor** [comment]
cant you check angel between A and new position and if is bigger than eg 5 degree, than call full pathfinder?

**Comment by raxtor** [comment]
I assume it could work like Follow me mod.. for chasing if do u understand me..

**Comment by raxtor** [comment]
Hi @Stephan-S 
I was playing with github version. Thx for circle stuck solution. After some observation I think many issues is caused by turn radius.
![fsScreen_2019_10_07_01_13_38](https://user-images.githubusercontent.com/831266/66277914-baeebf00-e8a4-11e9-8b00-445d369488ba.png)

Green course is working for tractor without trailer but when you attach trailer, turn radius is deferent to calculated route because back joint point has left/right degree limitation. Unfortunately every tractor has this value different with every trailer. If LUADOC has no function how calculate it, than turn radius has to be reduced about some universal value.

**Comment by Stephan-S** [comment]
I actually use the provided minTurnRadius for the attached trailer, if it is provided. Otherwise, the default value is used.
You can check the used radius by typing this into the console:
printVariable g_currentMission.controlledVehicle.ad.pf.minTurnRadius
or if you have also CP installed:
printVehicleVariable ad.pf.minTurnRadius

Since my pathfinder algorithm uses 45° steps between segments, this value is set to half the minTurnRadius of the implement + 1.5m to have some leeway.
If none is provided, it defaults back to the 6m, which was previously used for all vehicles.

So you could check if it uses the default value for your overloader or if one is provided.

Best regards,
Stephan

**Comment by Martinnygaard** [comment]
When using the Krampe Bandit 750 trailer for unloading, the script returns 10m.
This is still needing very much baby sitting. since updating to 1.0.6.1 it has not completet one unload without help.
Most of the times, it ends up cgoing in circles trying to reach a point it cannot reach.

> 2019-10-07 10:55 g_currentMission.controlledVehicle.ad.pf.minTurnRadius - nil
2019-10-07 10:55 Printing g_currentMission.controlledVehicle.ad.pf.minTurnRadius (number), depth 1
2019-10-07 10:55 g_currentMission.controlledVehicle.ad.pf.minTurnRadius: 10

**Comment by Stephan-S** [comment]
@Martinnygaard 
Thank you for your feedback, but please update to the latest sources. Your input from the release version 1.0.6.1 is currently not really helpul since I already updated the behaviour quite a lot :-)

**Comment by raxtor** [comment]
Well.. 
![fsScreen_2019_10_07_14_26_47](https://user-images.githubusercontent.com/831266/66311360-8fa0ba00-e90e-11e9-94ab-1c83a131decc.png)
tested for QuadTrack and Puma 135 tractor... I got circle stuck immediately after combine calling. (btw time for recorvery form circle stuck can be half of current state) and it was finished by crash to combine...
Both vehicles had set nil value when pathfinding active. So default 6m is in fact 12m + 1,5m? What I remember, CP is using usually 10-12+ So question is why there is so many circle stucks now when u didnt change it..? And for me it doesnt seems like 13,5m turn radius...

**Comment by raxtor** [comment]
Short test of github version...
![fsScreen_2019_10_11_00_56_34](https://user-images.githubusercontent.com/831266/66612960-0dd9b680-ebc4-11e9-847d-8746dd3bfe97.png)
Now there is opposite problem than before. 1st bin filled correctly than stucked between bins... in addtion there is some top cover flickering around unloading/loading waypoints. Top cover is also closed after combine unloading. In real life this top cover is not used on field.
Active unloading has still issues... almost disappeared circle stuck but pathfinding is unreliable and is leading through crop. Biggest troubles I had with unfinished headlands where pathfinder has big trouble find path and combine Now there is opposite problem than before. 1st bin filled correctly than stucked between bins... in addtion there is some top cover flickering around unloading/loading waypoints. Top cover is also closed after combine unloading. In real life this top cover is not used on field.
Active unloading has still issues... almost disappeared circle stuck but pathfinding is unreliable and is leading through crop. Biggest troubles I had with unfinished headlands where pathfinder has big trouble find path and lineup tractor to combine combine.. Pre-call and chasing off...

**Comment by raxtor** [comment]
@Stephan-S 
Maybe I found problem which is causing troubles for unloading to Wilson trailer...
See picture above, unloading point is set to Farm1 from unknown reason when unloader reached unloading point. I have set original unload point for every round again. When I set correct unloading point, than tractor is able fill 2nd bin too.

**Comment by raxtor** [comment]
@Stephan-S 
1.0.6.5
can you take a look at that problem above... 1st bin working well, after fullfilled 1st bin is unloading point changed and tractor is stucked...

anyway this version is big improvment for active unloading.. need longer time for testing, but it works almost good.

**Comment by raxtor** [comment]
![fsScreen_2019_11_02_22_16_38](https://user-images.githubusercontent.com/831266/68077117-8811eb00-fdbe-11e9-8153-759c880b6897.png)
1.0.6.6c github build - main issues update
- Following mode - tractor is following combine always in next 1 right row...
- Still problem with wilson trailer
- Still problem with unloading point for auger wagon

**Comment by raxtor** [comment]
btw I noticed some strange behaviour of combine pipes.. when multiple machines than all combines unfold their pipes in same time... I feel feel like unloader is confused which combine should be unloaded and sometimes leave original combine and going unload second one...
Also combine pipe is shortly unfolded when combine reached pre-call level
![fsScreen_2019_11_02_23_03_04](https://user-images.githubusercontent.com/831266/68077491-420c5580-fdc5-11e9-949d-e228f3d90f15.png)

**Comment by raxtor** [comment]
An now I got situation when tractor is stucked in circle and is not able go unload 1st combine.. (error msg - unloader is blocked). It seems like 2nd combine is blocking his path to 1st combine, but in fact path is free..
![fsScreen_2019_11_02_23_09_10](https://user-images.githubusercontent.com/831266/68077538-0625c000-fdc6-11e9-9e94-82ec73b8534a.png)

**Comment by Stephan-S** [comment]
So everything is working now?

**Comment by raxtor** [comment]
It seems no other report troubles, so probably yes...

**Comment by Stephan-S** [comment]
You know that I'm always working on this right?
Btw: you can view how the pathfinder works by typing ADsetDebugLevel into the console and then activating the 'Editor mode' on the unloade before he calculates a path.
Might be interesting for you to see how the pathfinder works.
Just toggle the edit mode again to let him drive on.

**Comment by Iwan1803** [comment]
Da kein weiteres Feedback -> geschlossen.
Bitte mit der neuesten Version erneut versuchen und bei Bedarf ein neues Issue erstellen, danke!
_Since no further feedback -> closed.
Please try with the latest version and create a new issue if necessary, thanks!_
